android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedNotificationScheduler.java

```diff
@@ -0,0 +1,119 @@
+package versioned.host.exp.exponent.modules.universal.notifications;
+
+import android.content.Context;
+import android.os.Bundle;
+import android.os.ResultReceiver;
+
+import org.unimodules.core.Promise;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import expo.modules.notifications.notifications.NotificationSerializer;
+import expo.modules.notifications.notifications.interfaces.NotificationTrigger;
+import expo.modules.notifications.notifications.model.NotificationContent;
+import expo.modules.notifications.notifications.model.NotificationRequest;
+import expo.modules.notifications.notifications.scheduling.NotificationScheduler;
+import expo.modules.notifications.notifications.service.ExpoNotificationSchedulerService;
+import host.exp.exponent.kernel.ExperienceId;
+
+public class ScopedNotificationScheduler extends NotificationScheduler {
+  @FunctionalInterface
+  private interface ScheduleNotificationFunction {
+    void invoke(String identifier, final Promise promise);
+  }
+
+  private final ExperienceId mExperienceId;
+
+  public ScopedNotificationScheduler(Context context, ExperienceId experienceId) {
+    super(context);
+    mExperienceId = experienceId;
+  }
+
+  @Override
+  protected NotificationRequest createNotificationRequest(String identifier, NotificationContent content, NotificationTrigger notificationTrigger) {
+    return new ScopedNotificationRequest(identifier, content, notificationTrigger, mExperienceId);
+  }
+
+  @Override
+  protected Collection<Bundle> serializeScheduledNotificationRequests(Collection<NotificationRequest> requests) {
+    Collection<Bundle> serializedRequests = new ArrayList<>(requests.size());
+    for (NotificationRequest request : requests) {
+      if (ScopedNotificationsUtils.shouldHandleNotification(request, mExperienceId)) {
+        serializedRequests.add(NotificationSerializer.toBundle(request));
+      }
+    }
+    return serializedRequests;
+  }
+
+  @Override
+  public void cancelScheduledNotificationAsync(String identifier, final Promise promise) {
+    ScheduleNotificationFunction baseFunction = super::cancelScheduledNotificationAsync;
+    ExpoNotificationSchedulerService.enqueueFetch(getContext(), identifier, new ResultReceiver(HANDLER) {
+      @Override
+      protected void onReceiveResult(int resultCode, Bundle resultData) {
+        super.onReceiveResult(resultCode, resultData);
+        if (resultCode == ExpoNotificationSchedulerService.SUCCESS_CODE) {
+          NotificationRequest request = resultData.getParcelable(ExpoNotificationSchedulerService.NOTIFICATION_REQUESTS_KEY);
+          if (request == null || !ScopedNotificationsUtils.shouldHandleNotification(request, mExperienceId)) {
+            promise.resolve(null);
+          }
+
+          baseFunction.invoke(identifier, promise);
```

Comment:
Hm… is there some specific reason to adding the notion of `ScheduleNotificationFunction`? Why not `ScopedNotificationScheduler.super.cancelScheduledNotificationAsync()`?

Isn't this a problem that the `onReceiveResult` now holds a strong reference to `ScopedNotificationScheduler` or is it playing too safe? If method conflict is difficult to solve we could add a
```java
private void doCancelScheduledNotificationAsync(String identifier, Promise promise) {
  super.cancelScheduledNotificationAsync(identifier, promise);
}
```
and call it on the weak reference.

Reply:
So, maybe it's an overengineering 😅I thought it would be a nice solution, but I didn't consider adding a new method :D So, I'll change it ;)

Reply:
And I'll also change the similar construction in the `PresentationModule` 😅